### PR TITLE
Explicitly set the node placement values when updating.

### PR DESCRIPTION
**Pull request by maya-r**
Merging doesn't have the right behaviour: we want the values in the
DaemonSet to be identical to the ones in the CR, even if those values
are empty.

Note that there's no unit testing for it, as the fake client appears
to suffer from the same Update merge behaviour, and setting to an
empty value fails.

### CODE REVIEW

**Comment by awels** [comment]
Let me understand what is failing if you update the CR to not include placement after it did have placement it doesn't update?

**Comment by maya-r** [comment]
if the DaemonSet Spec.Template.Spec.Nodeselector is: kubernetes.io/arch: amd64
and the HPP CR Spec.Workload.NodeSelector is empty

The result of merging the two is still NodeSelector: kubernetes.io/arch: amd64

File: pkg/controller/hostpathprovisioner/daemonset.go [hunk 1]
```diff
@@ -191,3 +195,13 @@ func createDaemonSetObject(cr *hostpathprovisionerv1.HostPathProvisioner, provis
...
 }
+
+// updateDaemonSetObject returns a DaemonSet with values set from CR.
+func updateDaemonSetObject(cr *hostpathprovisionerv1.HostPathProvisioner, orig runtime.Object) runtime.Object {
+	newDaemonSet := orig.(*appsv1.DaemonSet)
+	newDaemonSet.Spec.Template.Spec.NodeSelector = cr.Spec.Workloads.NodeSelector
+	newDaemonSet.Spec.Template.Spec.Tolerations = cr.Spec.Workloads.Tolerations
+	newDaemonSet.Spec.Template.Spec.Affinity = cr.Spec.Workloads.Affinity
+
+	return newDaemonSet.DeepCopyObject()
```
Comment (mhenriks) [inline pkg/controller/hostpathprovisioner/daemonset.go:206]: Are you sure you want to update `orig` and return a copy of it?  I think you would either want to only mutate the copy or mutate orig and return nothing.

**Comment by awels** [comment]
#95 fixes the broken reconcile loop by looking at the correct configuration object.

**Comment by aglitke** [comment]
Missing tests?

**Comment by awels** [comment]
No the tests are in hostpath provisioner. But as I said #95 fixes the problem for all the objects.

**Comment by awels** [comment]
We should be able to close this.

**Pull request closed without merging**